### ISSUE TITLE
update Twitter integration to support retrieving e-mail addresses

### ISSUE THREAD

**Issue by benlavalley** [meteor/meteor#4224]
While they have not updated their official documentation yet, if you request that your app be enabled for it, Twitter supports retrieving a user's e-mail address. 

Once support is added to your app by the Twitter support team, the address can be pulled when hitting verify_credentials with the include_email=true parameter. 

If the users' email address is confirmed on Twitter's end, it should be returned. 

I'm working on updating my own code to try and pull this now, but after looking at the core meteor Twitter package it looks like it would be pretty simple to add. I didn't spend more than an hour or so working on it, but I did try to pull down Meteor's core Twitter package and tweaking twitter_server.js to support pulling this in myself -- linked here for reference:

https://github.com/meteor/meteor/blob/556c0e28e94b9351cbf0b28e80a71a4e35f1362a/packages/twitter/twitter_server.js

I unfortunately got some sort of OATH already loaded error with my feeble attempts at removing the existing twitter package with meteor remove twitter and then creating my own custom package. I had to move on to coding other things on my app  and I don't know when I can circle back to this one, but maybe someone out there can look into this further.

**Comment by glasser** [comment]
Thanks for the feature request! We welcome discussions about how to make Meteor better. If you haven't yet, check out our wiki page about [feature requests on GitHub](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor#feature-requests).

Interesting; do you have any sources that document this?  When I do a Google web search for `"verify_credentials" "include_email" twitter` it only gets this very issue.

**Comment by shilman** [comment]
Twitter community forum has a big thread on this. @nibbl you might find the latest comment in the thread useful:

https://twittercommunity.com/t/how-to-get-email-from-twitter-user-using-oauthtokens/558/201

I'll take a stab at this in my Meteor app next week and will post here and/or issue a PR if I can figure anything out.

**Comment by nibbl** [comment]
I actually got this working two weeks ago. This is a function I run that is part of a larger function that updates Meteor user information from Twitter/Google/Facebook:

I'm not a programmer -- you're about to see an awful set of code below -- please don't hold it against me :) 

The key bit is in Twit.get.

```javascript
var twitteremailfunction = function (userid,callback) {
                    var wrappedTwitterEmailInsert = Meteor.bindEnvironment(function(userid,emailaddress) {
                        console.log ('running db update for user id of '+userid+' and email address of '+emailaddress);
                        if (!(Meteor.users.findOne({_id:userid,emails:{$exists:true}}))) {
                            Meteor.users.update({_id:userid},{$set:{emails:[]}});
                        }
                        Meteor.users.update({_id: userid}, {$set: {'emails.0.address': emailaddress,'emails.0.verified': false}}, function (error, _id) {
                            if (!(isEmpty(error))) {
                                if (typeof error.message !== 'undefined') {
                                    //logger.error('Error 9 setting username for '+userid+' could not set to '+newusername+'error is : '+error.message);
                                    callback(error.message);
                                } else {
                                    //logger.error('Error 9 setting username for '+userid+' could not set to '+newusername+' error is : '+error);
                                    callback(error);
                                }
                            } else {
                                if (typeof _id !== 'undefined') {
                                    if (_id !== 0) {
                                        logger.info('successfully set emailaddress to ', emailaddress, 'for user id ', userid);
                                        callback();
                                    } else {
                                        //logger.error('Unknown Error 10 setting username for '+userid+' could not set to '+newusername);
                                        callback('Unknown Error 10 setting emailaddress for '+userid+' could not set to '+emailaddress);
                                    }
                                } else {
                                    //logger.error('Unknown Error 11 setting username for '+userid+' could not set to '+newusername);
                                    callback('Unknown Error 11 setting emailaddress for '+userid+' could not set to '+emailaddress);
                                }
                            }
                        });


                    }, "Failed to update email address for twitter user");

                    if (!(isEmpty(Meteor.users.findOne({_id: userid}).services))) {
                        if (!(isEmpty(Meteor.users.findOne({_id: userid}).services.twitter))) {

                            var conf = JSON.parse(Assets.getText('twitter.json'));
                            var accesstokenkey = Meteor.users.findOne({_id: userid}).services.twitter.accessToken;
                            var accesstokensecret = Meteor.users.findOne({_id: userid}).services.twitter.accessTokenSecret;

                            Twit = new TwitMaker
                            ({
                                consumer_key: conf.consumer.key,
                                consumer_secret: conf.consumer.secret,
                                access_token_key: accesstokenkey,
                                access_token: accesstokenkey,
                                access_token_secret: accesstokensecret
                            });

                            Twit.get('account/verify_credentials',
                                {
                                    include_email:true
                                },
                                function(err, data, response) {
                                    console.log(data);
                                    // get email from twitter
                                    var email = data.email;
                                    console.log('this is the email that came from twitter: '+email);
                                    if(!(isEmpty(email))) {
                                        wrappedTwitterEmailInsert(userid,email);
                                    }else{
                                        logger.info('no email address returned from twitter');
                                        callback();
                                    }
                                }
                            );
                        }
                    }
                };
```

**Comment by abernix** [comment]
Thanks for opening this feature request.  It looks like this feature request was overlooked when a PR was submitted by someone else to solve exactly this problem.

This was fixed earlier this year by #6443 and is available in Meteor 1.3.1 and newer.  Closing! :)
